inc/Search/AbstractIndex.php

```diff
@@ -0,0 +1,403 @@
+<?php
+
+namespace dokuwiki\Search;
+
+use dokuwiki\Utf8;
+
+/**
+ * Abstract Class DokuWiki Index
+ *
+ * @license    GPL 2 (http://www.gnu.org/licenses/gpl.html)
+ * @author     Andreas Gohr <andi@splitbrain.org>
+ * @author Tom N Harris <tnharris@whoopdedo.org>
+ */
+abstract class AbstractIndex
+{
+    /* pages will be marked as deleted in page.idx */
+    const INDEX_MARK_DELETED = '#deleted:';
+
+    /** @var array $pidCache Cache for getPID() */
+    protected static $pidCache = array();
+
+    /**
+     * AbstractIndex constructor
+     * extended classes should be Singleton, prevent direct object creation
+     */
+    protected function __construct() {}
+
+    /**
+     * Clean a name of a key for use as a file name.
+     *
+     * Romanizes non-latin characters, then strips away anything that's
+     * not a letter, number, or underscore.
+     *
+     * @author Tom N Harris <tnharris@whoopdedo.org>
+     *
+     * @param string $name
+     * @return string
+     */
+    protected function cleanName($name)
+    {
+        $name = Utf8\Clean::romanize(trim((string)$name));
+        $name = preg_replace('#[ \./\\:-]+#', '_', $name);
+        $name = preg_replace('/[^A-Za-z0-9_]/', '', $name);
+        return strtolower($name);
+    }
+
+    /**
+     * Get the numeric PID of a page
+     *
+     * Warning: The page may not exist in the filesystem.
+     *
+     * @param string $page The page to get the PID for
+     * @return int|false  The page id on success, false when not found in page.idx
+     */
+    public function getPID($page)
+    {
+        // return PID when it is in the cache
+        // avoid expensive addIndexKey operation for the most recently
+        // requested pages by using a cache
+        if (isset(static::$pidCache[$page])) return static::$pidCache[$page];
```

Comment:
I think making the page cache static might be wrong. This makes all child classes use the same single cache array but children might want to use a different "page". Eg. a media index might store media IDs instead of pages. Or am I missing something?

Reply:
AFAIK DokuWIki does not have **media** index feature. We have FulltextIndex and MetadataIndex, both use only pageid as the static variable is named as `$pidCache`.

Reply:
@splitbrain I had wrong understanding about how the pidCache work between Indexer and other index classes. 
The changes in a32da6d does make Indexer pass a pid to MetadataIndex and FulltextIndex, and these subclass does not need to call `getPID()` to get numeric page id to be indexed. It might also mean the pidCache is unnecessary?

Reply:
The idea behind having this abstract indexing class, is that we could have additional indexes in the future and there the PIDs might refer to something else, not the common page.idx but a media.idx or whatever. A static variable will be the same for all instances of the class and it's descendants. So in that case a static cache here would be incorrect. ..

Reply:
Ah I stand corrected. I experimented a bit and an inheriting class can actually redefine a static member, so it's cache is separate. TIL.